### ISSUE TITLE
haplo runs to a halt with a large phased VCF file to annotate

### ISSUE THREAD

**Issue by ipstone** [Ensembl/ensembl-vep#840]
Hello, 

I am running to annotate a very large phased VCF file (~500GB) with haplo, but it runs to a stall after about 2 days, with only 140k lines annotated (memory usage is around 32GB - which is the allocated max memory for the computation node.

Might this issue be the result of 'norma' haplo run, due to the large input vcf used?
(I have searched the issues, and noticed: https://github.com/Ensembl/ensembl-vep/issues/497 it indicates installing bioperl-ext, might help with memory/performance issue with haplo - might this help)?

I tried to look for the info whether haplo can run in parallele/multi-threaded, but didn't find such info (is it possible?). If no parallel option, probably I will need to split the large vcf into smaller ones and run them through haplo?

Your help and suggestions are greatly appreciated.  Here is my haplo install information:
```
 haplo --helpPossible precedence issue with control flow operator at /lila/data/riazlab/lib/miniconda3/lib/site_perl/5.26.2/Bio/DB/IndexedBase.pm line 805.
#-------------#
# HAPLOSAURUS #
#-------------#

Versions:  ensembl              : 100.7e964b7
  ensembl-funcgen      : 100.f0c3948
  ensembl-io           : 100.f87ae4f
  ensembl-variation    : 100.1074e16
  ensembl-vep          : 100.1
Help: dev@ensembl.org , helpdesk@ensembl.org
Twitter: @ensembl

Usage:
perl haplo.pl [--cache|--offline|--database] [arguments]

Basic options
=============

--help                 Display this message and quit

-i | --input_file      Input file
-o | --output_file     Output file
--force_overwrite      Force overwriting of output file
--species [species]    Species to use [default: "human"]
```

**Comment by at7** [comment]
Hello,

haplo doesn't have an option to be run in parallel/multi-threaded mode. You need to split your input in order to run haplo in parallel. Be careful to split your input in a way that variants that overlap the same protein coding gene are not separated.
I would also recommend that you use the --offline option and install the cache files for looking up transcript information.
Finally, haplosaurus can run faster and in some cases with less memory using the bioperl-ext package. So if after dividing up your input and running in offline mode, you still encounter memory issues, maybe you could try installing the bioperl-ext package.

Best wishes,
Anja

**Comment by ipstone** [comment]
Thanks Anja for the all the info. I got my problem solved by splitting the file. I also tried to install bioperl-ext, but on my system it doesn't compile, the error part is as the following. I am guessing it could be the issue of the linker ld setup with the conda and linuxbrew environment - if you have tips on how to get around this, it will be also very helpful as well.

Thanks again and feel free to close this issue, as my questions were addressed, the question on bioperl-ext is quite tangential.

Thanks!


```           ^~~~~~~~~~~~~~~~~~~
           bp_sw_Exon_alloc
Align.xs: In function 'XS_Bio__Ext__Align__Histogram_extreme_fit_p':
Align.xs:1487:11: warning: implicit declaration of function 'bp_sw_ExtremeValueP'; did you mean 'bp_sw_Exon_alloc'? [-Wimplicit-function-declaration]
  RETVAL = bp_sw_ExtremeValueP(his,mu,lambda,score);
           ^~~~~~~~~~~~~~~~~~~
           bp_sw_Exon_alloc
rm -f ../blib/arch/auto/Bio/Ext/Align/Align.so
/data/lab/lib/miniconda3/bin/..//bin/x86_64-conda_cos6-linux-gnu-gcc  -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--disable-new-dtags -Wl,--gc-sections -Wl,-rpath,/data/lab/lib/miniconda3/lib -Wl,-rpath-link,/data/lab/lib/miniconda3/lib -L/data/lab/lib/miniconda3/lib --sysroot /data/lab/lib/miniconda3/bin/..//x86_64-conda_cos6-linux-gnu/sysroot -fstack-protector-strong  Align.o  -o ../blib/arch/auto/Bio/Ext/Align/Align.so libs/libsw.a \
   -lm   \

/cluster/data/lab/lib/miniconda3/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: libs/libsw.a(alnrange.o): relocation R_X86_64_PC32 against symbol `bp_sw_compare_AlnRange_start' can not be used when making a shared object; recompile with -fPIC
/cluster/data/lab/lib/miniconda3/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:463: ../blib/arch/auto/Bio/Ext/Align/Align.so] Error 1
make[1]: Leaving directory '/cluster/home/ipstone/Workbench/bioperl-ext/Bio/Ext/Align'
make: *** [Makefile:636: subdirs] Error 2
```

**Comment by at7** [comment]
Hi Isaac,
I cannot help with the bioperl-ext question and will close this issue. But please reopen or create a new issue if you have any other questions.
Best wishes,
Anja
